src/core/lwip/tcp_in.c

```diff
@@ -947,6 +947,15 @@ static void ack_partial_or_whole_segment(struct tcp_pcb *pcb, u32_t ackno, struc
     struct tcp_seg *whole_seg_to_ack;
     while ((*seg) != NULL && TCP_SEQ_GT(ackno, (*seg)->seqno)) {
         if (TCP_SEQ_LT(ackno, (*seg)->seqno + TCP_SEGLEN((*seg)))) {
+            if ((*seg)->tcp_flags & TCP_FIN) {
```

Comment:
1. the fact we retransmit all the data - even if it was partially ACKED is OK and the comment is good but imho should be documented in redmine as well as a known issue.

Suggestion for the issue I'd add:
Splitting the last FIN-bearing segment into two (data-only and FIN-only segments) can allow partial ACK to safely remove the data portion while preserving the FIN portion. requires a deeper refactor of how segments are created and queued, but will make the session more efficient.

2. if there's a window update (e.g. - due to congestion) for some reason, can retransmitting all the data be a problem that will stall us?